modules/pam_userdb/pam_userdb.c

```diff
@@ -287,7 +287,7 @@ user_lookup (pam_handle_t *pamh, const char *database, const char *cryptmode,
 	    } else {
 #ifdef HAVE_CRYPT_R
 	      struct crypt_data *cdata = NULL;
-	      cdata = malloc(sizeof(*cdata));
+	      cdata = calloc(1, sizeof(*cdata));
```

Comment:
Not part of the change introduced here, but the `cdata` variable is not overridden after its usage. If this does not belong on this PR, it should be fixed later on. Otherwise an extra commit might be useful here.

And talking about it, I think the allocation can be skipped because `cdata` could reside in stack, which removes the error handling if allocation fails.

Reply:
Added a commit to zero out the data.

Do we want to use the stack for `struct crypt_data`, since its size seems to be 32768 bytes?


It's currently used once on the stack here: https://github.com/linux-pam/linux-pam/blob/cf8c50f2c76fd29c018a2220d243f41801aaadb1/modules/pam_pwhistory/opasswd.c#L130
and otherwise on the heap.

Reply:
Okay, knowing about its size now I am definitely against having it on stack. Especially since PAM might be used in threads which should not be that excessive with their stacks.

At a later point in time (not part of this PR IMHO since it's all about clearing data) the stack usage in pam_pwhistory should disappear instead.

Reply:
Added commit to move the allocation to the heap.